File: AoC2022/day23.cpp
#include <fstream>
#include <vector>
#include <stack>
#include <iostream>
#include <list>
#include <unordered_set>
#include <unordered_map>
#include <map>

using namespace std;

struct pair_hash {
    inline std::size_t operator()(const std::pair<int,int> & v) const {
        return v.first*31+v.second;
    }
};

void draw_map(vector<pair<int,int>> &elves) {
    int minx = INT32_MAX, miny = INT32_MAX, maxx = INT32_MIN, maxy = INT32_MIN;
    unordered_set<pair<int,int>, pair_hash> cur_map;
    for (auto const& elf: elves)
    {
        minx = min(elf.first, minx);
        miny = min(elf.second, miny);
        maxx = max(elf.first, maxx);
        maxy = max(elf.second, maxy);
        cur_map.insert(elf);
    }

    for (int y = miny; y <= maxy; y++){
        for (int x = minx; x <= maxx; x++)
        {
            if (cur_map.count({x,y})) cout << '#';
            else cout << '.';
        }
        cout << endl;
    }
    cout << endl;
}

int main()
{
    ifstream file("day23.txt");
    vector<string> lines;
    while (file.good())
    {
        string line;
        getline(file, line);
        lines.push_back(line);
    }
    file.close();

    vector<pair<int, int>> elves;
    unordered_set<pair<int, int>, pair_hash> cur_map;

    for (int y = 0; y < lines.size(); y++)
    {
        for (int x = 0; x < lines[y].length(); x++)
        {
            if (lines[y][x] == '#')
            {
                elves.push_back({x, y});
                cur_map.insert({x,y});
            }
        }
    }

    long round = 0;
    long elves_moving = 1;
    while (elves_moving > 0)
    {
        unordered_map<pair<int, int>, vector<pair<int, int> *>, pair_hash> new_map;
        elves_moving = 0;

        for (auto &elf : elves)
        {
            pair<int, int> ne = {elf.first + 1, elf.second - 1};
            pair<int, int> n = {elf.first, elf.second - 1};
            pair<int, int> nw = {elf.first - 1, elf.second - 1};
            pair<int, int> w = {elf.first - 1, elf.second};
            pair<int, int> sw = {elf.first - 1, elf.second + 1};
            pair<int, int> s = {elf.first, elf.second + 1};
            pair<int, int> se = {elf.first + 1, elf.second + 1};
            pair<int, int> e = {elf.first + 1, elf.second};

            pair<int, int> newelf = elf;

            int has_ne =cur_map.count(ne);
            int has_n = cur_map.count(n);
            int has_nw =cur_map.count(nw);
            int has_w = cur_map.count(w);
            int has_sw =cur_map.count(sw);
            int has_s = cur_map.count(s);
            int has_se =cur_map.count(se);
            int has_e = cur_map.count(e);

            if (has_ne || has_n || has_nw || has_w ||
                has_sw || has_s || has_se || has_e)
            {
                for (int i = 0; i < 4; i++)
                {
                    int to_check = (round + i) % 4;
                    if (to_check == 0)
                    {
                        if (!(has_ne || has_n || has_nw))
                        {
                            newelf = n;
                            break;
                        }
                    }
                    else if (to_check == 1)
                    {
                        if (!(has_sw || has_s || has_se))
                        {
                            newelf = s;break;
                        }
                    }
                    else if (to_check == 2)
                    {
                        if (!(has_sw || has_w || has_nw))
                        {
                            newelf = w;break;
                        }
                    }
                    else if (to_check == 3)
                    {
                        if (!(has_se || has_e || has_ne))
                        {
                            newelf = e;break;
                        }
                    }
                }
            }
            if (newelf != elf)
                elves_moving++;

            // Store pointer to elf at new position
            if (new_map.count(newelf) == 0)
            {
                new_map.insert({newelf, {}});
            }
            new_map[newelf].push_back(&elf);
        }

        // Update elves that moved to an empty position
        for (auto const &p : new_map)
        {
            if (p.second.size() == 1)
            {
                p.second[0]->first = p.first.first;
                p.second[0]->second = p.first.second;
            }
        }

        // Create new map
        cur_map.clear();
        for (auto const& elf: elves)
        {
            cur_map.insert(elf);
        }

        //draw_map(elves);

        round++;
        if (round == 10) {
            int minx = INT32_MAX, miny = INT32_MAX, maxx = INT32_MIN, maxy = INT32_MIN;
            for (auto const& elf: elves)
            {
                minx = min(elf.first, minx);
                miny = min(elf.second, miny);
                maxx = max(elf.first, maxx);
                maxy = max(elf.second, maxy);
            }
            cout << "Part 1: " << (maxx - minx + 1) * (maxy - miny + 1) - elves.size() << endl;
        }
    }
    cout << "Part 2: " << round << endl;
}

File: AoC2022/day23.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;

namespace AoC2021
{
    class Day202223
    {
        class Elf
        {
            public Elf(int x, int y)
            {
                this.x = x;
                this.y = y;
            }
            public int x;
            public int y;
        }

        void drawmap(List<Elf> elves)
        {
            int minx = elves[0].x;
            int miny = elves[0].y;
            int maxx = elves[0].x;
            int maxy = elves[0].y;
            var map = new HashSet<(int, int)>();
            foreach (var elf in elves)
            {
                minx = Math.Min(minx, elf.x);
                maxx = Math.Max(maxx, elf.x);
                miny = Math.Min(miny, elf.y);
                maxy = Math.Max(maxy, elf.y);
                map.Add((elf.x, elf.y));
            }

            for (int y = miny; y <= maxy; y++)
            {
                for (int x = minx; x <= maxx; x++)
                {
                    if (map.Contains((x, y)))
                    {
                        Console.Write('#');
                    }
                    else
                    {
                        Console.Write('.');
                    }
                }
                Console.WriteLine();
            }
            Console.WriteLine();
        }

        internal void Solve()
        {
            var input = System.IO.File.ReadAllLines("input23.txt");
            var currentmap = new HashSet<(int, int)>();

            var elves = new List<Elf>();
            for (int y = 0; y < input.Length; y++)
            {
                for (int x = 0; x < input[y].Length; x++)
                {
                    if (input[y][x] == '#')
                    {
                        var newelf = new Elf(x, y);
                        elves.Add(newelf);
                        currentmap.Add((x, y));//, new List<Elf>() { newelf });
                    }
                }
            }

            int loopcount = 10;
            int roundcount = 0;
            while(true)
            {
                int nomove = 0;
                var newmap = new Dictionary<(int, int), List<Elf>>();
                foreach (var elf in elves)
                {
                    var w = (elf.x - 1, elf.y);
                    var nw = (elf.x - 1, elf.y - 1);
                    var n = (elf.x, elf.y - 1);
                    var ne = (elf.x + 1, elf.y - 1);
                    var e = (elf.x + 1, elf.y);
                    var se = (elf.x + 1, elf.y + 1);
                    var s = (elf.x, elf.y + 1);
                    var sw = (elf.x - 1, elf.y + 1);
                    var elfmoved = false;

                    if (!(currentmap.Contains(nw) || currentmap.Contains(n) || currentmap.Contains(ne) || currentmap.Contains(e) ||
                          currentmap.Contains(w) || currentmap.Contains(sw) || currentmap.Contains(s) || currentmap.Contains(se)))
                    {
                        elfmoved = false;
                    }
                    else
                    {
                        for (int c = 0; c < 4; c++)
                        {
                            int tocheck = (roundcount + c) % 4;
                            if (tocheck == 0)
                            {
                                if (!(currentmap.Contains(nw) || currentmap.Contains(n) || currentmap.Contains(ne)))
                                {
                                    if (!newmap.ContainsKey(n))
                                    {
                                        newmap.Add(n, new List<Elf>());
                                    }
                                    newmap[n].Add(elf);
                                    elfmoved = true;
                                    break;
                                }
                            }
                            else if (tocheck == 1)
                            {
                                if (!(currentmap.Contains(sw) || currentmap.Contains(s) || currentmap.Contains(se)))
                                {
                                    if (!newmap.ContainsKey(s))
                                    {
                                        newmap.Add(s, new List<Elf>());
                                    }
                                    newmap[s].Add(elf);
                                    elfmoved = true;
                                    break;
                                }
                            }
                            else if (tocheck == 2)
                            {
                                if (!(currentmap.Contains(sw) || currentmap.Contains(w) || currentmap.Contains(nw)))
                                {
                                    if (!newmap.ContainsKey(w))
                                    {
                                        newmap.Add(w, new List<Elf>());
                                    }
                                    newmap[w].Add(elf);
                                    elfmoved = true;
                                    break;
                                }
                            }
                            else if (tocheck == 3)
                            {
                                if (!(currentmap.Contains(se) || currentmap.Contains(e) || currentmap.Contains(ne)))
                                {
                                    if (!newmap.ContainsKey(e))
                                    {
                                        newmap.Add(e, new List<Elf>());
                                    }
                                    newmap[e].Add(elf);
                                    elfmoved = true;
                                    break;
                                }
                            }
                        }
                    }
                    if (elfmoved == false)
                    {
                        if (!newmap.ContainsKey((elf.x, elf.y)))
                        {
                            newmap.Add((elf.x, elf.y), new List<Elf>());
                        }
                        newmap[(elf.x, elf.y)].Add(elf);
                        nomove++;
                    }
                }

                foreach (var p in newmap)
                {
                    if (p.Value.Count == 1)
                    {
                        foreach (var e in p.Value)
                        {
                            e.x = p.Key.Item1;
                            e.y = p.Key.Item2;
                        }
                    }
                }

                currentmap.Clear();
                foreach (var e in elves)
                {
                    currentmap.Add((e.x, e.y));//, e);
                }



                // drawmap(elves);
                roundcount++;
                if (nomove == elves.Count)
                {
                    break;
                }
            }


            int minx = elves[0].x;
            int miny = elves[0].y;
            int maxx = elves[0].x;
            int maxy = elves[0].y;
            foreach (var elf in elves)
            {
                minx = Math.Min(minx, elf.x);
                maxx = Math.Max(maxx, elf.x);
                miny = Math.Min(miny, elf.y);
                maxy = Math.Max(maxy, elf.y);
            }

            Console.WriteLine($"Part 1: { (maxx - minx + 1) * (maxy - miny + 1) - elves.Count}");
            Console.WriteLine($"Part 2: {roundcount}");
        }


    }
}
